text/0000-scoped-take-2.md

```diff
@@ -0,0 +1,261 @@
+        F: FnOnce() -> T + Send + 'a,
+        T: Send + 'a;
+}
+```
+
+Like the original `thread::scoped` API, this allows for the child thread's
+closure to borrow data (with lifetime `'body`) from the parent thread. These
+borrows are bounded by the lifetime of the `Scope` value, and the implementation
+uses the `defer` method to add a callback that will join on (wait for completion
+of) the child thread on exit from the scope -- thus restoring memory safety.
+
+Note that, while previously one might return `JoinGuard`s outward to expand the
+scope of joining, the pattern here is reversed: you call `scope` at the
+outer-most scope, and then pass a reference to the `Scoped` value inward.
+
+Putting it all together, here's an example from TRPL, and a version adjust to the new API:
+
+```rust
+// using the thread::scoped API
+fn old_trpl_example() {
+    let data = Mutex::new(vec![1u32, 2, 3]);
+
+    let guards: Vec<_> = (0..2).iter().map(|_| {
+        thread::scoped(|| {
+            let mut data = data.lock().unwrap();
+            data[i] += 1;
+        })
+    }).collect();
+
+    // threads implicitly joined here, when `guards` goes out of scope and drops
+    // its contents
+}
+```
+
+```rust
+// using the proposed thread::scope API
+fn new_trpl_example() {
+    let data = Mutex::new(vec![1u32, 2, 3]);
+
+    thread::scope(|s| {
+        for i in 0..2 {
+            s.spawn(|| {
+                let mut data = data.lock().unwrap();
+                data[i] += 1;
+            });
+        }
+    })
+}
+```
+
+In the original version of the example, the join guards from the scoped threads
+were explicitly collected into a vector (to ensure that the joins did not happen
+too early).
+
+With the new version, by contrast, the scope is more clearly marked by an indent,
+all *all* joins within the scope automatically happen at the end of the block,
+making scoped threads feel more like a first-class control-flow construct.  In
+this RFC author's opinion, this actually winds up *clarifying* the semantics of
+scoped threads, and so may be a better API than the original `thread::scoped`.
+
+On a separate note, in practice one will usually not want to spawn full-fledged
+threads when doing data-parallel, fork-join style computations; instead you want
+to use a thread pool, work stealing, and higher-level combinators. One nice
+aspect of `Scope` is that crates providing such parallelism frameworks can
+easily hook into the API with their own means of spawning lightweight tasks;
+this should also facilitate a [decoupling of thread pools from the lifetimes
+bounding the tasks they are running](https://github.com/rust-lang/threadpool/issues/7).
+
+## Implementation
+
+Here's a sketch of the implementation, inspired in part by
+[@arielb1's ideas](https://gist.github.com/arielb1/5eb299a87546ce8829b3):
+
+```rust
+pub struct Scope<'a> {
+    dtors: RefCell<Option<DtorChain<'a>>>
+}
+
+struct DtorChain<'a> {
+    dtor: Box<FnBox() + 'a>,
```

Comment:
Unfortunately `Box<FnOnce()>` isn't callable because by-value `self` methods on trait objects are not yet callable. `FnBox` is a trait that has a `call_box` methods which takes `Box<Self>`, allowing it to be called. It's hoped that this will be removed one day though!